integrationtests/self/http_test.go

```diff
@@ -1068,4 +1069,118 @@ var _ = Describe("HTTP tests", func() {
 			"Unannounced": {"Surprise!"},
 		})))
 	})
+
+	It("graceful shutdown successfully with no timeout", func() {
+		// server will begin shutdown after receiving two requests
+		var (
+			fastChan = make(chan struct{})
+			slowChan = make(chan struct{})
+			wg       sync.WaitGroup
+		)
+
+		mux.HandleFunc("/fast", func(w http.ResponseWriter, r *http.Request) {
+			close(fastChan)
+			w.Write(PRDataLong)
+		})
+		mux.HandleFunc("/slow", func(w http.ResponseWriter, r *http.Request) {
+			close(slowChan)
+			ticker := time.NewTicker(time.Second)
+			defer ticker.Stop()
+			chunkSize := len(PRDataLong) / 10
+			for i := range 10 {
+				<-ticker.C
+				w.Write(PRDataLong[i*chunkSize : (i+1)*chunkSize])
+			}
+		})
+		// makes two requests, one fast and one slow, both are expected to finish successfully
+		wg.Add(1)
+		go func() {
+			resp, err := client.Get(fmt.Sprintf("https://localhost:%d/fast", port))
+			Expect(err).ToNot(HaveOccurred())
+			Expect(resp.StatusCode).To(Equal(200))
+			body, err := io.ReadAll(resp.Body)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(body).To(Equal(PRDataLong))
+			wg.Done()
+		}()
+		wg.Add(1)
+		go func() {
+			resp, err := client.Get(fmt.Sprintf("https://localhost:%d/slow", port))
+			Expect(err).ToNot(HaveOccurred())
+			Expect(resp.StatusCode).To(Equal(200))
+			body, err := io.ReadAll(resp.Body)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(body).To(Equal(PRDataLong))
+			wg.Done()
+		}()
+
+		wg.Add(1)
+		go func() {
+			<-fastChan
+			<-slowChan
+			err := server.CloseGracefully(context.Background())
+			Expect(err).ToNot(HaveOccurred())
+			wg.Done()
+		}()
+		wg.Wait()
+	})
+
+	It("graceful shutdown successfully with timeout", func() {
+		// server will begin shutdown after receiving two requests
+		var (
+			fastChan = make(chan struct{})
+			slowChan = make(chan struct{})
+			wg       sync.WaitGroup
```

Comment:
Avoid `sync.WaitGroup`. It might block indefinitely when something goes wrong, which is very annoying to debug. Instead, use a channel and `Eventually`, so you can actually fail the test.